File: app/src/main/java/com/prox/voicechanger/ui/activity/RecordActivity.java
package com.prox.voicechanger.ui.activity;

import static com.prox.voicechanger.VoiceChangerApp.TAG;
import static com.prox.voicechanger.ui.dialog.MoreOptionDialog.SELECT_AUDIO;
import static com.prox.voicechanger.ui.dialog.TextToVoiceDialog.IMPORT_TEXT;
import static com.prox.voicechanger.utils.PermissionUtils.REQUEST_PERMISSION;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.speech.tts.TextToSpeech;
import android.speech.tts.UtteranceProgressListener;
import android.util.Log;
import android.view.WindowManager;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.fragment.NavHostFragment;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

import com.prox.voicechanger.R;
import com.prox.voicechanger.VoiceChangerApp;
import com.prox.voicechanger.databinding.ActivityRecordBinding;
import com.prox.voicechanger.databinding.DialogLoading2Binding;
import com.prox.voicechanger.databinding.DialogRateBinding;
import com.prox.voicechanger.ui.dialog.LoadingDialog;
import com.prox.voicechanger.ui.dialog.RateDialog;
import com.prox.voicechanger.ui.dialog.TextToVoiceDialog;
import com.prox.voicechanger.utils.FileUtils;
import com.prox.voicechanger.utils.PermissionUtils;
import com.prox.voicechanger.viewmodel.FileVoiceViewModel;
import com.proxglobal.proxads.adsv2.callback.AdsCallback;

import java.io.File;
import java.util.HashMap;
import java.util.Locale;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class RecordActivity extends AppCompatActivity {
    public static final String IMPORT_TO_CHANGE_VOICE = "IMPORT_TO_CHANGE_VOICE";
    public static final String IMPORT_TEXT_TO_SPEECH = "IMPORT_TEXT_TO_SPEECH";
    private NavController navController;
    private AppBarConfiguration appBarConfiguration;
    private TextToSpeech mTts;
    private FileVoiceViewModel model;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Log.d(TAG, "RecordActivity: onCreate");
        super.onCreate(savedInstanceState);
        ActivityRecordBinding binding = ActivityRecordBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        model = new ViewModelProvider(this).get(FileVoiceViewModel.class);
        model.isExecuteText().observe(this, isExecute -> {
            if (isExecute){
                VoiceChangerApp.instance.showInterstitial(RecordActivity.this, "interstitial_text", new AdsCallback() {
                    @Override
                    public void onClosed() {
                        super.onClosed();
                        Intent goToChangeVoice = new Intent(RecordActivity.this, ChangeVoiceActivity.class);
                        goToChangeVoice.setAction(IMPORT_TEXT_TO_SPEECH);
                        goToChangeVoice.putExtra(ChangeVoiceActivity.PATH_FILE, FileUtils.getTempTextToSpeechFilePath(RecordActivity.this));
                        startActivity(goToChangeVoice);
                        overridePendingTransition(R.anim.anim_right_left_1, R.anim.anim_right_left_2);
                        Log.d(TAG, "RecordActivity: To ChangeVoiceActivity");
                    }

                    @Override
                    public void onError() {
                        super.onError();
                        Intent goToChangeVoice = new Intent(RecordActivity.this, ChangeVoiceActivity.class);
                        goToChangeVoice.setAction(IMPORT_TEXT_TO_SPEECH);
                        goToChangeVoice.putExtra(ChangeVoiceActivity.PATH_FILE, FileUtils.getTempTextToSpeechFilePath(RecordActivity.this));
                        startActivity(goToChangeVoice);
                        overridePendingTransition(R.anim.anim_right_left_1, R.anim.anim_right_left_2);
                        Log.d(TAG, "RecordActivity: To ChangeVoiceActivity");
                    }
                });
            }else {
                Toast.makeText(RecordActivity.this, R.string.process_error, Toast.LENGTH_SHORT).show();
            }
        });

        init();
    }

    @Override
    protected void onDestroy() {
        Log.d(TAG, "RecordActivity: onDestroy");
        appBarConfiguration = null;
        navController = null;
        if (mTts != null) {
            mTts.shutdown();
        }
        super.onDestroy();
    }

    @Override
    public boolean onSupportNavigateUp() {
        if (navController == null || appBarConfiguration == null) {
            return super.onSupportNavigateUp();
        } else {
            return NavigationUI.navigateUp(navController, appBarConfiguration)
                    || super.onSupportNavigateUp();
        }
    }

    private void init() {
        Log.d(TAG, "RecordActivity: init");

        getWindow().addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
        getWindow().setStatusBarColor(this.getResources().getColor(R.color.background_app));

        NavHostFragment navHostFragment = (NavHostFragment) getSupportFragmentManager()
                .findFragmentById(R.id.nav_host_record_activity);
        if (navHostFragment != null) {
            navController = navHostFragment.getNavController();
        } else {
            Log.d(TAG, "RecordActivity: navHostFragment null");
            recreate();
            return;
        }

        appBarConfiguration = new AppBarConfiguration.Builder(navController.getGraph()).build();

        PermissionUtils.checkPermission(this, this);

        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
        int open_app = preferences.getInt("open_app", 1);
        if (open_app >= 2){
            RateDialog dialog = new RateDialog(this, DialogRateBinding.inflate(getLayoutInflater()), () -> {});
            dialog.show();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_PERMISSION) {
            if (grantResults.length > 0
                    && grantResults[0] == PackageManager.PERMISSION_GRANTED
                    && grantResults[1] == PackageManager.PERMISSION_GRANTED
                    && grantResults[2] == PackageManager.PERMISSION_GRANTED) {
//                if (navController == null){
//                    Log.d(TAG, "RecordActivity: navController null");
//                }else {
//                    navController.navigate(R.id.action_recordFragment_to_stopRecordFragment);
//                    Log.d(TAG, "RecordActivity: To StopRecordFragment");
//                }
            } else {
                PermissionUtils.openDialogAccessAllFile(this);
            }
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_PERMISSION) {
            int record = ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO);
            int write = ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE);
            int read = ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE);
            if (record == PackageManager.PERMISSION_GRANTED
                    && write == PackageManager.PERMISSION_GRANTED
                    && read == PackageManager.PERMISSION_GRANTED) {
//                if (navController == null){
//                    Log.d(TAG, "RecordActivity: navController null");
//                }else {
//                    navController.navigate(R.id.action_recordFragment_to_stopRecordFragment);
//                    Log.d(TAG, "RecordActivity: To StopRecordFragment");
//                }
            } else {
                PermissionUtils.openDialogAccessAllFile(this);
            }
        } else if (requestCode == SELECT_AUDIO) {
            if (resultCode == Activity.RESULT_OK) {
                if (data == null) {
                    Log.d(TAG, "RecordActivity: data null");
                } else {
                    String filePath = FileUtils.getRealPath(this, data.getData());
                    if (filePath.isEmpty()) {
                        Log.d(TAG, "RecordActivity: filePath isEmpty");
                        Toast.makeText(this, R.string.file_not_exist, Toast.LENGTH_SHORT).show();
                    } else if (!(new File(filePath).exists())) {
                        Log.d(TAG, "RecordActivity: filePath not exists");
                        Toast.makeText(this, R.string.file_not_exist, Toast.LENGTH_SHORT).show();
                    } else {
                        Log.d(TAG, "RecordActivity: filePath " + filePath);

                        Intent goToChangeVoice = new Intent(this, ChangeVoiceActivity.class);
                        goToChangeVoice.setAction(IMPORT_TO_CHANGE_VOICE);
                        goToChangeVoice.putExtra(ChangeVoiceActivity.PATH_FILE, filePath);
                        startActivity(goToChangeVoice);
                        overridePendingTransition(R.anim.anim_right_left_1, R.anim.anim_right_left_2);
                        Log.d(TAG, "RecordActivity: To ChangeVoiceActivity");
                    }
                }
            } else if (resultCode == Activity.RESULT_CANCELED) {
                Toast.makeText(this, R.string.canceled, Toast.LENGTH_SHORT).show();
            }
        } else if (requestCode == IMPORT_TEXT) {
            VoiceChangerApp.instance.showInterstitial(RecordActivity.this, "interstitial_text", new AdsCallback() {
                @Override
                public void onClosed() {
                    super.onClosed();

                }

                @Override
                public void onError() {
                    super.onError();

                }
            });
            if (resultCode == TextToSpeech.Engine.CHECK_VOICE_DATA_PASS) {
                LoadingDialog dialog = new LoadingDialog(
                        this,
                        DialogLoading2Binding.inflate(getLayoutInflater())
                );
                dialog.show();

                mTts = new TextToSpeech(this, status -> {
                    if (status == TextToSpeech.SUCCESS) {
                        mTts.setOnUtteranceProgressListener(new UtteranceProgressListener() {
                            @Override
                            public void onStart(String s) {
                                Log.d(TAG, "onStart");
                            }

                            @Override
                            public void onDone(String s) {
                                Log.d(TAG, "onDone");
                                model.setExecuteText(true);
                                dialog.cancel();
                                Log.d(TAG, "RecordActivity: To ChangeVoiceActivity");
                            }

                            @Override
                            public void onError(String s) {
                                Log.d(TAG, "onError");
                                model.setExecuteText(false);
                                dialog.cancel();
                            }
                        });

                        mTts.setLanguage(Locale.US);

                        if (TextToVoiceDialog.textToSpeech.isEmpty()) {
                            Toast.makeText(this, R.string.process_error, Toast.LENGTH_SHORT).show();
                            return;
                        }

                        HashMap<String, String> params = new HashMap<>();
                        params.put(TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID, TextToVoiceDialog.textToSpeech);

                        Log.d(TAG, "textToSpeech: " + TextToVoiceDialog.textToSpeech);
                        mTts.synthesizeToFile(TextToVoiceDialog.textToSpeech, params, FileUtils.getTempTextToSpeechFilePath(this));
                    }
                });
            } else {
                Intent installIntent = new Intent();
                installIntent.setAction(TextToSpeech.Engine.ACTION_INSTALL_TTS_DATA);
                startActivity(installIntent);
                overridePendingTransition(R.anim.anim_right_left_1, R.anim.anim_right_left_2);
            }
        }
    }
}

File: app/src/main/java/com/prox/voicechanger/ui/activity/RecordActivity.kt
package com.prox.voicechanger.ui.activity

import android.Manifest
import com.prox.voicechanger.utils.FileUtils.Companion.getTempTextToSpeechFilePath
import androidx.navigation.ui.NavigationUI.navigateUp
import com.prox.voicechanger.utils.PermissionUtils.checkPermission
import com.prox.voicechanger.utils.PermissionUtils.openDialogAccessAllFile
import com.prox.voicechanger.utils.FileUtils.Companion.getRealPath
import dagger.hilt.android.AndroidEntryPoint
import androidx.appcompat.app.AppCompatActivity
import androidx.navigation.NavController
import androidx.navigation.ui.AppBarConfiguration
import android.speech.tts.TextToSpeech
import com.prox.voicechanger.viewmodel.FileVoiceViewModel
import android.os.Bundle
import com.prox.voicechanger.VoiceChangerApp
import androidx.lifecycle.ViewModelProvider
import android.content.Intent
import com.prox.voicechanger.R
import android.widget.Toast
import android.view.WindowManager
import androidx.navigation.fragment.NavHostFragment
import android.content.pm.PackageManager
import android.net.Uri
import androidx.core.content.ContextCompat
import com.prox.voicechanger.ui.dialog.MoreOptionDialog
import com.prox.voicechanger.ui.dialog.TextToVoiceDialog
import com.prox.voicechanger.ui.dialog.LoadingDialog
import android.speech.tts.TextToSpeech.OnInitListener
import android.speech.tts.UtteranceProgressListener
import android.util.Log
import com.bumptech.glide.Glide.init
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.prox.voicechanger.databinding.ActivityRecordBinding
import com.prox.voicechanger.databinding.DialogLoading2Binding
import com.prox.voicechanger.model.FileVoice
import com.prox.voicechanger.ui.dialog.OptionDialog
import com.prox.voicechanger.ui.dialog.OptionDialog.Companion.fileVoice
import com.prox.voicechanger.utils.PermissionUtils.REQUEST_PERMISSION
import kotlinx.coroutines.*
import java.io.File
import java.util.*

@AndroidEntryPoint
class RecordActivity : AppCompatActivity() {
    private var navController: NavController? = null
    private var appBarConfiguration: AppBarConfiguration? = null
    private var mTts: TextToSpeech? = null
    private var model: FileVoiceViewModel? = null
    private val activityScope = CoroutineScope(Job() + Dispatchers.Main)

    private lateinit var auth: FirebaseAuth
    private val db: FirebaseFirestore = FirebaseFirestore.getInstance()
    private var storage: FirebaseStorage? = null
    private var storageRef: StorageReference? = null
    private var riversRef: StorageReference? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        Log.d(VoiceChangerApp.TAG, "RecordActivity: onCreate")
        super.onCreate(savedInstanceState)
        val binding = ActivityRecordBinding.inflate(
            layoutInflater
        )
        setContentView(binding.root)
        model = ViewModelProvider(this).get(FileVoiceViewModel::class.java)
        model!!.isExecuteText().observe(this) { isExecute: Boolean ->
            if (isExecute) {
                val goToChangeVoice = Intent(this@RecordActivity, ChangeVoiceActivity::class.java)
                goToChangeVoice.action = IMPORT_TEXT_TO_SPEECH
                goToChangeVoice.putExtra(
                    ChangeVoiceActivity.PATH_FILE,
                    getTempTextToSpeechFilePath(this@RecordActivity)
                )
                startActivity(goToChangeVoice)
                overridePendingTransition(R.anim.anim_right_left_1, R.anim.anim_right_left_2)
                Log.d(VoiceChangerApp.TAG, "RecordActivity: To ChangeVoiceActivity")
            } else {
                Toast.makeText(this@RecordActivity, R.string.process_error, Toast.LENGTH_SHORT)
                    .show()
            }
        }
        init()

        storage = FirebaseStorage.getInstance()
        storageRef = storage!!.reference

        auth = FirebaseAuth.getInstance()
    }

    override fun onStart() {
        super.onStart()
        if (auth.currentUser != null) {
            activityScope.launch {
                val listFileVoice = model!!.getFileVoices().value

                for (fileVoice in listFileVoice!!) {
                    withContext(Dispatchers.IO) {
                        val docRef =
                            db.collection(auth.currentUser!!.uid).document(fileVoice!!.name)
                        docRef.get().addOnSuccessListener { document ->
                            if (document.data == null) {
                                val data = hashMapOf(
                                    "id" to fileVoice.id,
                                    "src" to fileVoice.src,
                                    "name" to fileVoice.name,
                                    "path" to fileVoice.path,
                                    "duration" to fileVoice.duration,
                                    "size" to fileVoice.size,
                                    "date" to fileVoice.date
                                )
                                db.collection(auth.currentUser!!.uid)
                                    .document(fileVoice.name)
                                    .set(data)

                                val file = Uri.fromFile(File(fileVoice.path))

                                riversRef =
                                    storageRef!!.child("${auth.currentUser!!.uid}/${fileVoice.name}")
                                riversRef!!.putFile(file)
                            }
                        }
                    }
                }
            }
        }
    }

    override fun onDestroy() {
        Log.d(VoiceChangerApp.TAG, "RecordActivity: onDestroy")
        appBarConfiguration = null
        navController = null
        if (mTts != null) {
            mTts!!.shutdown()
        }
        super.onDestroy()
    }

    override fun onSupportNavigateUp(): Boolean {
        return if (navController == null || appBarConfiguration == null) {
            super.onSupportNavigateUp()
        } else {
            (navigateUp(navController!!, appBarConfiguration!!)
                    || super.onSupportNavigateUp())
        }
    }

    private fun init() {
        Log.d(VoiceChangerApp.TAG, "RecordActivity: init")
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        window.statusBarColor = this.resources.getColor(R.color.background_app)
        val navHostFragment = supportFragmentManager
            .findFragmentById(R.id.nav_host_record_activity) as NavHostFragment?
        navController = if (navHostFragment != null) {
            navHostFragment.navController
        } else {
            Log.d(VoiceChangerApp.TAG, "RecordActivity: navHostFragment null")
            recreate()
            return
        }
        appBarConfiguration = AppBarConfiguration.Builder(navController!!.graph).build()
        checkPermission(this, this)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_PERMISSION) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED && grantResults[1] == PackageManager.PERMISSION_GRANTED && grantResults[2] == PackageManager.PERMISSION_GRANTED) {
//                if (navController == null){
//                    Log.d(TAG, "RecordActivity: navController null");
//                }else {
//                    navController.navigate(R.id.action_recordFragment_to_stopRecordFragment);
//                    Log.d(TAG, "RecordActivity: To StopRecordFragment");
//                }
            } else {
                openDialogAccessAllFile(this)
            }
        }
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_PERMISSION) {
            val record = ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO)
            val write =
                ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
            val read =
                ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE)
            if (record == PackageManager.PERMISSION_GRANTED && write == PackageManager.PERMISSION_GRANTED && read == PackageManager.PERMISSION_GRANTED) {
//                if (navController == null){
//                    Log.d(TAG, "RecordActivity: navController null");
//                }else {
//                    navController.navigate(R.id.action_recordFragment_to_stopRecordFragment);
//                    Log.d(TAG, "RecordActivity: To StopRecordFragment");
//                }
            } else {
                openDialogAccessAllFile(this)
            }
        } else if (requestCode == MoreOptionDialog.SELECT_AUDIO) {
            if (resultCode == RESULT_OK) {
                if (data == null) {
                    Log.d(VoiceChangerApp.TAG, "RecordActivity: data null")
                } else {
                    val filePath = getRealPath(this, data.data)
                    if (filePath!!.isEmpty()) {
                        Log.d(VoiceChangerApp.TAG, "RecordActivity: filePath isEmpty")
                        Toast.makeText(this, R.string.file_not_exist, Toast.LENGTH_SHORT).show()
                    } else if (!File(filePath).exists()) {
                        Log.d(VoiceChangerApp.TAG, "RecordActivity: filePath not exists")
                        Toast.makeText(this, R.string.file_not_exist, Toast.LENGTH_SHORT).show()
                    } else {
                        Log.d(VoiceChangerApp.TAG, "RecordActivity: filePath $filePath")
                        val goToChangeVoice = Intent(this, ChangeVoiceActivity::class.java)
                        goToChangeVoice.action = IMPORT_TO_CHANGE_VOICE
                        goToChangeVoice.putExtra(ChangeVoiceActivity.PATH_FILE, filePath)
                        startActivity(goToChangeVoice)
                        overridePendingTransition(
                            R.anim.anim_right_left_1,
                            R.anim.anim_right_left_2
                        )
                        Log.d(VoiceChangerApp.TAG, "RecordActivity: To ChangeVoiceActivity")
                    }
                }
            } else if (resultCode == RESULT_CANCELED) {
                Toast.makeText(this, R.string.canceled, Toast.LENGTH_SHORT).show()
            }
        } else if (requestCode == TextToVoiceDialog.IMPORT_TEXT) {
            if (resultCode == TextToSpeech.Engine.CHECK_VOICE_DATA_PASS) {
                val dialog = LoadingDialog(
                    this,
                    DialogLoading2Binding.inflate(layoutInflater)
                )
                dialog.show()
                mTts = TextToSpeech(this, OnInitListener label@{ status: Int ->
                    if (status == TextToSpeech.SUCCESS) {
                        mTts!!.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
                            override fun onStart(s: String) {
                                Log.d(VoiceChangerApp.TAG, "onStart")
                            }

                            override fun onDone(s: String) {
                                Log.d(VoiceChangerApp.TAG, "onDone")
                                model!!.setExecuteText(true)
                                dialog.cancel()
                                Log.d(VoiceChangerApp.TAG, "RecordActivity: To ChangeVoiceActivity")
                            }

                            @Deprecated("Deprecated in Java")
                            override fun onError(s: String) {
                                Log.d(VoiceChangerApp.TAG, "onError")
                                model!!.setExecuteText(false)
                                dialog.cancel()
                            }
                        })
                        mTts!!.language = Locale(TextToVoiceDialog.code_language!!)
                        if (TextToVoiceDialog.textToSpeech!!.isEmpty()) {
                            Toast.makeText(this, R.string.process_error, Toast.LENGTH_SHORT).show()
                            return@label
                        }
                        val params = HashMap<String, String?>()
                        params[TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID] =
                            TextToVoiceDialog.textToSpeech
                        Log.d(
                            VoiceChangerApp.TAG,
                            "textToSpeech: " + TextToVoiceDialog.textToSpeech
                        )
                        mTts!!.synthesizeToFile(
                            TextToVoiceDialog.textToSpeech,
                            params,
                            getTempTextToSpeechFilePath(this)
                        )
                    }
                })
            } else {
                val installIntent = Intent()
                installIntent.action = TextToSpeech.Engine.ACTION_INSTALL_TTS_DATA
                startActivity(installIntent)
                overridePendingTransition(R.anim.anim_right_left_1, R.anim.anim_right_left_2)
            }
        }
    }

    companion object {
        const val IMPORT_TO_CHANGE_VOICE = "IMPORT_TO_CHANGE_VOICE"
        const val IMPORT_TEXT_TO_SPEECH = "IMPORT_TEXT_TO_SPEECH"
    }
}